### PR TITLE
feat(loki sink): add support for structured metadata

### PR DESCRIPTION
**Pull request by maxboone**
See issue: https://github.com/vectordotdev/vector/issues/18876

Initial attempt on this, but seeking some feedback to wrap up.

**How can we deal with the change in protocol, as structured metadata is part of a newer protocol and this PR overrides the current proto for Loki.** 

More precisely, see the first two commits of this PR for the proto changes:
https://github.com/vectordotdev/vector/pull/20576/files/20285e7544fe57312f4d170dfda9e2e6b5672e31

**How extensive do we need tests to be, do we want ({ labels, no labels, dynamic labels} x { structured metadata, no structured metadata, dynamic structured metadata}) or is the addition of tests that only cover the new code sufficient?**

### CODE REVIEW

**Comment by tobz** [comment]
As far as the question around changes to the Protocol Buffers definition -- and I haven't look at the changes yet -- I _think_ it should be fine to use them as-is, so long as the changes as all backwards-compatible.

The addition of new fields to existing message types should be fine, as older servers should just ignore them, as far as I understand. I'm not sure how most gRPC server implementations handle requests for unknown gRPC _services_, though, if we're dealing with new RPC calls.

File: src/sinks/loki/config.rs [hunk 1]
```diff
@@ -66,6 +66,15 @@ pub struct LokiConfig {
     #[serde(default = "crate::serde::default_false")]
     pub remove_label_fields: bool,
 
+    /// Fields to add to structured metadata.
```
Comment (tobz) [inline src/sinks/loki/config.rs:69]:
```suggestion
    /// Structured metadata that is attached to each batch of events.
    ///
    /// Both keys and values are templateable, which enables you to attach dynamic structured metadata to events.
    ///
    /// Valid metadata keys include `*`, and prefixes ending with `*`, to allow for the expansion of
    /// objects into multiple metadata entries. This follows the same logic as [Label expansion][label_expansion].
    ///
    /// [label_expansion]: https://vector.dev/docs/reference/configuration/sinks/loki/#label-expansion
```

File: src/sinks/loki/sink.rs [hunk 2]
```diff
@@ -776,14 +919,85 @@ mod tests {
...
+            Template::try_from("cluster_name").unwrap(),
+            Template::try_from("static_cluster_name").unwrap(),
+        );
+
+        let mut encoder = EventEncoder {
+            key_partitioner: KeyPartitioner::new(None),
+            transformer: Default::default(),
+            encoder: Encoder::<()>::new(JsonSerializerConfig::default().build().into()),
+            labels: HashMap::default(),
+            structured_metadata: Option::None,
```
Comment (tobz) [inline src/sinks/loki/sink.rs:943]: Seems like you should actually be setting this to the `structured_metadata` field you've constructed above?

**Review by tobz — CHANGES_REQUESTED** [review]
Overall, it seems like the Protocol Buffers changes themselves should be safe and backwards-compatible, so it mostly boils down to documentation change cleanup and fixing some of the bugs that seem to be present.

**Comment by kimxogus** [comment]
Hi, is there any update in this PR? I'm waiting for this feature

**Comment by maxboone** [comment]
> Hi, is there any update in this PR? I'm waiting for this feature

My bad, I had a lot of other work to attend to. Will pick this up Thursday.

**Comment by devkoriel** [comment]
Hi there, it's great to see the progress on this PR for adding support for structured metadata in the Loki sink! The addition of structured metadata is a significant enhancement that will greatly benefit users.

I understand there are some concerns regarding the changes in protocol and the extent of testing required. While the changes in Protocol Buffers definitions seem to be safe and backwards-compatible, it is crucial to ensure thorough testing to cover various scenarios like labels, structured metadata, and dynamic structured metadata.

Addressing the feedback provided by tobz regarding documentation and code cleanup will further improve the overall quality of the implementation. Making the necessary adjustments in the code and documentation as suggested will help in achieving a more cohesive and maintainable solution.

Given the importance of this feature and the work put into it, I strongly encourage the team to review and merge this PR as soon as possible. This feature is eagerly awaited by users and will undoubtedly enhance the capabilities of the Loki sink in Vector. Let's work together to finalize the remaining tasks and get this valuable enhancement merged promptly. Thank you for your hard work on this PR!

**Comment by maxboone** [comment]
@tobz thanks a ton for the feedback, I think I processed everything but this is a first real rust contribution for me so apologies for it being messy here and there. 

I kept the implementation as similar to the labels as I saw possible but that probably breaks any DRY criteria as it's basically the same code done over. Is it preferred if I refactor these into a single function e.g. `convertTemplatedMaptoVec` and call that from the respective functions?

**Comment by maxboone** [comment]
@tobz I'll pick up the things that turn up from the CI tomorrow morning, if there's anything else lmk!

**Comment by maxboone** [comment]
@tobz any news here, please lmk if I can do anything!

**Comment by tobz** [comment]
@maxboone Apologies. I swore I had previously approved this, but it seems not!

LGTM overall, but it still needs a changelog entry. You can see the README for the changelog directory -- https://github.com/vectordotdev/vector/tree/master/changelog.d -- which explains what is required.

**Review by tobz — APPROVED** [review]

**Comment by maxboone** [comment]
> LGTM overall, but it still needs a changelog entry.

@tobz Will do, thanks a ton for the review!

**Comment by maxboone** [comment]
@tobz added the changelog entry, should be good to go :)

**Comment by maxboone** [comment]
@pront @tobz I see some CI failing, is there something I can change / update (or extend the tests and run integration tests on my end)?

I haven't tested it against a local instance of Loki recently, if I need to set that up to ensure it works properly lmk!

**Comment by jszwedko** [comment]
> @pront @tobz I see some CI failing, is there something I can change / update (or extend the tests and run integration tests on my end)?
> 
> I haven't tested it against a local instance of Loki recently, if I need to set that up to ensure it works properly lmk!

It looks like the integration tests are failing to compile. You should be able to get the same errors locally by running `make check-clippy`. You can also run the integration tests locally via `make test-integration-loki`.

**Comment by maxboone** [comment]
> It looks like the integration tests are failing to compile. You should be able to get the same errors locally by running `make check-clippy`. You can also run the integration tests locally via `make test-integration-loki`.

Will do, will ping again if something pops up :)

**Comment by maxboone** [comment]
I see some messages popping up from running clippy and the integration tests here. I really thought I had fixed those but reality seems different. I'll fix this in the morning (CEST) and poke again.

**Comment by jszwedko** [comment]
> I see some messages popping up from running clippy and the integration tests here. I really thought I had fixed those but reality seems different. I'll fix this in the morning (CEST) and poke again.
> 
> It's mostly just the protobuf body that has changed and thus the unit test failing.

np! Happy to review again afterwards.

**Comment by maxboone** [comment]
@jszwedko integration tests ran correctly with the latest changes, mainly added `#[serde(default)]` to `structured_metadata` in the LokiConfig, as the tests were failing on only the configuration and I personally would want to avoid adding a breaking change to existing configurations. 

Anyways, as far as I can replicate, the CI should succeed now!

**Comment by jszwedko** [comment]
Great, thanks @maxboone ! Let me trigger CI here.

**Comment by kimxogus** [comment]
👍

**Comment by miadabrin** [comment]
@maxboone I think it just needs a docs regeneration using `make generate-component-docs` so that the tests pass

**Comment by maxboone** [comment]
> @maxboone I think it just needs a docs regeneration using `make generate-component-docs` so that the tests pass

good call! will do in a minute

**Comment by kimxogus** [comment]
@maxboone Could you run `make generate-component-docs`?

**Comment by jszwedko** [comment]
I was in the area so I regenerated them.

**Review by jhgilbert — APPROVED** [review]

**Comment by maxboone** [comment]
> good call! will do in a minute

Yeah, unfortunately WSL2 decided to drop out that afternoon after an non-downgradeable update. But it's working again - thanks @jszwedko for generating the docs <3

**Pull request merged**